### ISSUE TITLE
WSL needs a pure launcher, decoupling invocation from any Windows Terminal handling (aka passthrough)

### ISSUE THREAD

**Issue by mintty** [microsoft/WSL#9117]
In microsoft/terminal#1173 there is a discussion about ConPTY passthrough mode for Windows Terminal, in order to support pure VT-style terminal applications to connect to applications.
However, in the context of WSL, the issue is a bit misplaced there as it belongs in the pure WSL domain.
The current launcher wsl.exe makes an unfortunate bundle between actually launching WSL, and providing any supposed connection to Windows Terminal, including conhost compatibility. These are two distinct functions and there should be a decoupled WSL launcher which can run in a terminal (whether remote, or a local application) and contains none of the Windows Terminal functionality at all in the first place. This would resolve WSL connectivity issues for a number of scenarios.

**Comment by o-sdn-o** [comment]
As far as I know, wsl.exe has nothing to do with Windows Terminal, it just utilize the Windows Console API for I/O and buffer size tracking.  wsl.exe uses a Console API Server which specified at startup.

**Comment by o-sdn-o** [comment]
In other words, in order to eliminate the dependence on Windows Terminal, you just need to implement the so-called Console API Server, and reference it in the thread attributes when wsl.exe is launched, just like Window Terminal itself does.

**Comment by mintty** [comment]
Maybe I misinterpreted the current scenario by incorrectly attributing it to Windows Terminal.
What I mean to address is that wsl.exe implements some terminal functionality, captures escape sequences and handles them in its own way, blocks other sequences, imposes its own idea of terminal size etc. It should have a transparent mode in which is does not do all that but simply transfers all controls between WSL and the client.

**Comment by o-sdn-o** [comment]
wsl.exe works in absolutely transparent mode. The behavior you describe applies to ConPTY, not wsl.exe.

**Comment by o-sdn-o** [comment]
ConPTY intercepts all I/O from wsl.exe and replaces it entirely with its own.

**Comment by mintty** [comment]
OK, consider this example: I run wsl.exe from a (non-Windows) terminal, in WSL I send ESC[c to the terminal, and the response is ESC[1;0c, fixed, without communicating that sequence with the client. If wsl.exe is not the culprit for this, how does ConPTY come into its way in this scenario and how do I get rid of it?

**Comment by o-sdn-o** [comment]
In order to receive input wsl.exe at startup subscribes on input events from condrv using something like AttachConsole+ReadConsoleInput. Conpty is responsible for handling these api calls. In other direction, wsl.exe outputs their output stream using WriteConsoleOutput, which is also the ConPTY territory.

**Comment by o-sdn-o** [comment]
You can play with it in terminals that do not use conpty approach. I know the only one - vtm, I am working on it. May be other exist on windows platform, idk.

**Comment by o-sdn-o** [comment]
Briefly, the wsl.exe launch sequence is as follows.  The terminal launches its Console API Server, the wsl.exe startup attributes specifies the reference to the running Console Server, and wsl.exe is launched.  The connection between the server and wsl.exe is provided by condrv belonging to the kernel. The core problem is that the condrv interface is documented only via refrence implementation at the Windows Terminal project.

**Comment by o-sdn-o** [comment]
You can run `vtm -r term wsl` in terminal you like. In that case vtm replaces ConPTY and you can check pure wsl.exe behavior. Moreover, you can compile a debug build. In debug mode, all calls to the Windows Console API are logged (you need to start vtm server separately with `vtm -s` and standalone `vtm` client in another console window).

**Comment by o-sdn-o** [comment]
Here is an important point, vtm does not pass the output from wsl.exe beyond itself and your terminal will not see almost anything from wsl.exe. vtm behaves like a full-fledged terminal and uses an external terminal only as a canvas.

**Comment by o-sdn-o** [comment]
Of course, I don't know wsl.exe internals for sure, and I, like you, look at it only from the outside, and I will be glad if at least something from the above clarifies wsl.exe's role in the chain.

**Comment by mintty** [comment]
That's github:netxs-group/vtm apparently. Thanks for the hint, maybe some useful handling can be derived from there.

**Comment by o-sdn-o** [comment]
> OK, consider this example: I run wsl.exe from a (non-Windows) terminal, in WSL I send ESC[c to the terminal, and the response is ESC[1;0c, fixed, without communicating that sequence with the client.

If I send ESC[c to the terminal from wsl session, vtm receives exactly the same ESC[c from wsl.exe

wsl.exe activity log:
```
6380:  pty: consrv: WriteConsole
6380:       raw write emulation
6380:       UTF-8: \e[c
```

https://user-images.githubusercontent.com/11535558/200134487-e33eab18-32b0-412f-b67e-441834c180ce.mp4
